### ISSUE TITLE
which --geno file to input from angsd?

### ISSUE THREAD

**Issue by devonorourke** [fgvieira/ngsLD#1]
Hi,

I was under the impression that the `.glf.gz` file would the the appropriate file to specify for the `--geno` argument when following ngsLD's [example usage](https://github.com/fgvieira/ngsLD#usage). However, I keep generating an error that appears to suggest that my input .glf file format is unacceptable for ngsLD to operate.

The .glf.gz file was produced with angsd, running the `-doGlf 3` option to generate a binary beagle file (see [here](http://www.popgen.dk/angsd/index.php/Genotype_Likelihoods#Output_genotype_likelihoods) for angsd documentation).

If for example, I provide the following code:
```
ngsLD \
--geno POP1.glf.gz \
--n_ind 7 \
--n_sites $NumSITES \
--pos POP1.glf.pos.gz \
--out POP1.ld
```

I get the following error:
```
==> GZIP input file (not BINARY)
> Reading data from file...
> Header found! Skipping line...

=======
WARNING: [read_geno]  header found but not on first line. Is this an error?
=======

^?<B5>^?^^<C0><9B>/n> Header found! Skipping line...

```
The file actually generates a whole bunch of error lines, which seem like it's attempting to read a binary file and perhaps that's the whole problem - does *ngsLD* want an unzipped non-binary glf file? 

Perhaps there's a way within angsd to convert the binary beagle file back into a more useful format for ngsLD to function. Or perhaps there is a way for ngsLD to read binary beagle-formatted files that I'm unclear about?

Thanks very much!

**Comment by fgvieira** [comment]
Hi Devon,

ngsLD assumes that only text files (like beagle) are gziped, while binary files (e.g. *.geno.gz) are not.
just try unziping the geno.gz file...

cheers,

**Comment by devonorourke** [comment]
Thanks for the help. 
A bit more context: the .glf file was created in angsd (v.0.918) using the following command:
```
angsd -P 24 \
-b POP1_bamlist.txt \
-ref myreference.fa -anc myreference.fa \
-out POP1 \
-uniqueOnly 0 -remove_bads 1 -only_proper_pairs 0 -trim 0 -C 50 -baq 1 -minMapQ 20 -minInd 4 -setMinDepthInd 3 -nInd 7 -SNP_pval 1e-3 -minMaf 0.05 \
-doDepth 1 -doCounts 1 -domajorminor 1 -domaf 1 -gl 1 -doGlf 3 -doSaf 2 -fold 1 \
-indF POP1.indF  \
```

The output of these arguments included two gzipped files `POP1.glf.gz` and `POP1.glf.pos.gz`.  

Using your advice, I unzipped those files and tried running the following command:
```
ngsLD \
--pos POP1.glf.pos \
--geno POP1.glf \
--n_ind 7 \
--n_sites 493592 \
--out POP1
```

It appears that the program was able to identify the file format correctly (binary), but it doesn't seem to like the `-doGlf 3` output from angsd format. This was the error thrown immediately after executing the command:
```
==> BINARY input file (always lkl)
> Reading data from file...

=====
ERROR: [read_geno] NaN found! Is the file format correct?
=====

	: Numerical argument out of domain
```

I thought that perhaps something was wrong with my .glf files; that perhaps I've run something incorrectly in angsd on the front end, but I doubt it - the position file contains all the expected sites of interest, the .saf file was then used to run `realSFS` in angsd (and worked), and those same arguments were subsequently used to also identify thetas later on. I don't think there is a problem with the data.

My only other thought is that there is something wrong about the format of the `-doGlf 3` option I selected within angsd when generating this .glf file. I wonder two things:
1. Am I mistaken that all of the standard output formats from angsd's `-doGlf` likelihood files are accepted for ngsLD? Or are there some exceptions?
2. If they are, I wonder if you could adopt a similar "angsd-style" flag that allows the user to specify which input format it is, rather than having to gzip a binary. In other words, something where within the ngsLD command you can just type `--glf 3` or `--glf 2` or `--glf 1` and the program would know that you're specifying a certain angsd output format. Maybe that raises too many complications.

Thanks!

**Comment by fgvieira** [comment]
Hi Devon, 

if you use `-doGlf 3` option, then you are working with genotype likelihoods in log scale. In this case, you need to specify `--probs --log_scale`

1. Most standard output formats from angsd are supported by ngsLD, but you need to specify the `--probs --log_scale` options accordingly. For example, `-doGlf` gives you genotype likelihoods, so you need to specify `--probs`. If you used `-doGlf 2`, then they are not log-scaled and no need for more options. But if you use `-doGlf 3`, then they are log-scaled and you need to specify the `--log_scale` option.
2. yes, I could do that, but then would loose compatibility with other programs. The basic idea is that with called genotypes no need for no extra options, genot likelihoods or post prob add `--probs`, and if in log add `--log_scale`. Briefly:
    1. `-doGlf 2` => `--probs`
    2. `-doGlf 3` => `--probs --log_scale` (with unzipped `geno.gz`)
    3. `-doGeno 2` => no extra option
    4. `-doGeno 32` => `--probs` (with unzipped `geno.gz`)

hope it is clear now.
cheers,

PS - `--log_scale` actually assumes `--probs`, so if you specify `--log_scale` it automatically assumes the `--probs`

**Comment by devonorourke** [comment]
Wonderful - thanks for the explanation regarding the expected flags to pass with this specific angsd glf file. I think what was confusing me was my literal interpretation of the [angsd documentation](http://www.popgen.dk/angsd/index.php/Genotype_Likelihoods#Output_genotype_likelihoods) regarding the .glf ouputs: it appeared to me that options `-dGglf 1` and `-doGlf4` were the only log-formatted outputs. Further, I couldn't find anywhere within the [Beagle-specific documentation](http://www.popgen.dk/angsd/index.php/Genotype_Likelihoods#Beagle_format) that suggested that the `-doGlf 3` output would be log scaled, just that it would be a binary version of `-doGlf 2`.

But you've solved the problem - the program ran without error!

Thanks for helping me through the process. Looking forward to using it on all my data now.

Cheers,
Devon

**Comment by fgvieira** [comment]
No problem, glad it works..
cheers,

**Comment by martabe** [comment]
Hi!

Just a comment on the specified option to be used with ANGSD `-doGlf 2`:
>        1. `-doGlf 2` => `--probs`

My input file is formatted as shown on the [ANGSD page](http://popgen.dk/angsd/index.php/Genotype_Likelihoods#Beagle_format): a header, tab-separated columns, and three probabilities per sample.
By using the `--probs` option I get the error
`wrong GENO file format. Less fields than expected!`
while without it everything goes fine.

Is this intended? Have I got the wrong file format? 

Thank you,
Marta

**Comment by fgvieira** [comment]
Hi Marta,

how many individuals and sites do you have?
how lines and columns do you have on your beagle file?

**Comment by martabe** [comment]
Hi!

I have 70 individuals and 766653 sites.

To count lines:
```
gzip -dc myinput.beagle.gz | wc -l
766654              # includes the header
``` 

To count columns (there is maybe a better solution for this, but I'm not super-skilled):
```
gzip -dc myinput.beagle.gz | head > tmp_beagle
awk '{print NF}' tmp_beagle | sort -nu | tail -n 1
213                   # includes marker, allele1 and allele2 columns
```

**Comment by fgvieira** [comment]
Can you send me the command line you used, plus what is printed on the screen?

**Comment by martabe** [comment]
I run the job on a cluster, so I get the STDERR and STDOUT in a file. STDOUT is empty.

This is the command I use:
```
ngsLD --geno myinput.beagle.gz --pos myinput.pos --n_ind 70 --n_sites 766653 --min_maf 0.10 --max_kb_dist 2000000 --out myoutput --n_threads 2
```

And the STDERR content:
```
==> Input Arguments:
	geno: /fake/path/myinput.beagle.gz
	probs: true
	log_scale: false
	n_ind: 70
	n_sites: 766653
	pos: /fake/path/myinput.pos
	max_kb_dist (kb): 200000
	max_snp_dist: 0
	min_maf: 0.100000
	call_geno: false
	N_thresh: 0.000000
	call_thresh: 0.000000
	rnd_sample: 1.000000
	seed: 1566307306
	extend_out: false
	out: /fake/path/myoutput
	n_threads: 2
	verbose: 1
	version: 1.0.0 (Aug 20 2019 @ 10:38:54)

==> GZIP input file (not BINARY)
> Reading data from file...
> Header found! Skipping line...

=====
ERROR: [read_geno] wrong GENO file format. Less fields than expected!
=====

	: Success
```

**Comment by fgvieira** [comment]
Does your `.pos` file have 766653 lines?

If so, can you try making a smaller `.beagle.gz` file (e.g. the first 2 or 3 lines) and send it to me together with the `.pos` file?

**Comment by martabe** [comment]
Yes, the .pos file has 766653 lines (there is no header in the file).

I had to rename myinput.pos into myinput.pos.txt to attach it here.
[myinput.pos.txt](https://github.com/fgvieira/ngsLD/files/3520974/myinput.pos.txt)

[myinput.beagle.gz](https://github.com/fgvieira/ngsLD/files/3520968/myinput.beagle.gz)

**Comment by fgvieira** [comment]
I just tried your files and they worked fine. Can you try updating your `ngsLD` to the latest version?

**Comment by martabe** [comment]
I'm sorry, I just realised that when I wrote the command in the comment above I forgot the `--probs` flag.

> ```
> ngsLD --geno myinput.beagle.gz --pos myinput.pos --n_ind 70 --n_sites 766653 --min_maf 0.10 --max_kb_dist 2000000 --out myoutput --n_threads 2
> ```

So just to be clear, if I add `--probs` as mentioned in your comment above (16/11/2017) I get the error, otherwise everything goes fine.

I just pulled the repository and compiled again and I get the same error by using the `--probs` flag.
Any idea what else I can check?

(BTW, thanks a lot for the help!)

**Comment by fgvieira** [comment]
Well, it seems there are some "weird" characters on your beagle file. How did you generate it? Was it directly from ANGSD, or did you manipulate the file in any way? Did you use Linux?

**Comment by martabe** [comment]
Hi,

I did not generate the file using ANGSD indeed. I used their example to build a genotype likelihood file with the same format. My own file comes from a "back-transformation" of the PL values in a GATK vcf file into GL values. I do this using a python script I wrote. And yes, I work on a Linux machine.

The symbols you see in the file are just numbers in exponential notation. I can round everything to a few digits in order not to have it, if you think that might be the problem. If so, do you have an idea of how many digits would be appropriate to keep?

**Comment by fgvieira** [comment]
Hi Marta,

it is not the exponential notation, that I think it works fine.
I found (e.g.) on the second line, between columns 4 and 5 some kind of weird (invisible) character that messed up the reading of the GL. Can you try:

```
gunzip myinput.beagle.gz
dos2unix myinput.beagle
gzip myinput.beagle
```

**Comment by martabe** [comment]
I converted the test file and now it works, thanks a lot! :)
I'm now running ngsLD on the original file (after converting it). It hasn't crashed yet.

I don't know why I have such weird characters in the file though; I've never used it on a dos machine.
Can I ask you how you've found them? I will try to fix my script in order to solve the problem.
Thanks so much again!

**Comment by fgvieira** [comment]
Glad it is working now!
I only found them because, when debugging the code, I saw that those two columns were not being properly split. But no idea why...

I haven't tried running `ngsLD` with GL from GATK. Can you let me know how it goes? If there are any errors/warnings, but also how the results look like; and specially if you end up doing a comparison with ANGSD! :)

**Comment by martabe** [comment]
Mmh, columns four and five are produced in the same way as the next ones... I'll see if I can spot the bug. Thanks for the hint though.

I'll let you know how it goes (I was hoping the job would finish in a few hours to give you the answer but it's taking some time. I could probably allocate some more threads and memory though). I'm not sure I will do a comparison with ANGSD GLs, but in case I will surely report to you.

Thank you again!
Marta

**Comment by martabe** [comment]
Hi Filipe!

It seems like the results I got are reasonable! I don't get any error/warning.
The LD does not get very low for some of the chromosomes, but that's expected for the samples we have (it's hybrids between two species of petunia, and we know we have lack of recombination).

I attach a plot of the LD decay (I modified the plotting code).
![decay_chromosome pdf](https://user-images.githubusercontent.com/25846389/64964607-b03f9c00-d89b-11e9-865c-af5d4c43757b.png)

I'm not planning to make a comparison with ANGSD GLs for now.

Thanks a lot for the help,
Marta

**Comment by fgvieira** [comment]
Glad to see it worked!
Do you think you can share the code to convert from GATK to GLs? I think it could be a nice addition to ngsLD... :)

**Comment by martabe** [comment]
Hi!

Sure I can share the code. Just be aware that I'm not a real programmer and this is a very very very quick and dirty script, ie:
- I did not figure out why I get some weird characters in the output
- so far I only used it on GATK vcfs which have passed through a vcftools filtering step as well (this shouldn't make a difference though)
- the output is comma-separated, while I think Beagle format is tab-separated (which is why I run the output through `sed -i "s/,/\\t/g"` afterwards)
- it possibly needs a long time and a lot of memory to run
- missing PL values are encoded as "0,0,0" by GATK, but I had some troubles getting this accepted by some other software down the pipeline, so I re-code them as "0.333333,0.333333,0.333333"
- individual names get recoded into "IndX" where X goes from 0 to the total number of individuals. A list of the new names gets printed to STDOUT

The formula I use consists in taking each of the three PL values (let's call them PL1, PL2, PL3) that characterise a sample at a certain position and calculating the "unscaledGL" ( 10^(-PL/10) ) for each of them. Then I take each of the unscaledGLs and divide it by the sum of the unscaledGLs to have them range between 0 and 1.

I attach the script anyway.

Also, vcftools has a --BEAGLE-PL option which might do the same (never checked the source code though).

Marta

[gatkPLtobeagleGL.py.txt](https://github.com/fgvieira/ngsLD/files/3651768/gatkPLtobeagleGL.py.txt)
